cpp/include/cudf/lists/detail/gather.cuh

```diff
@@ -0,0 +1,268 @@
+ * Thinking generally, this means that to produce a gather_map for level N+1, we need to use the
+ * offsets from level N and the "base" offsets from level N-1. So we are always carrying along
+ * one extra buffer of these "base" offsets which keeps our memory usage well controlled.
+ *
+ * @code{.pseudo}
+ * Example:
+ *
+ * "column of lists of lists of ints"
+ * {
+ *   {
+ *      {2, 3}, {4, 5}
+ *   },
+ *   {
+ *      {6, 7, 8}, {9, 10, 11}, {12, 13, 14}
+ *   },
+ *   {
+ *      {15, 16}, {17, 18}, {17, 18}, {17, 18}, {17, 18}
+ *   }
+ * }
+ *
+ * List<List<int32_t>>:
+ * Length : 3
+ * Offsets : 0, 2, 5, 10
+ * Children :
+ *    List<int32_t>:
+ *    Length : 10
+ *    Offsets : 0, 2, 4, 7, 10, 13, 15, 17, 19, 21, 23
+ *       Children :
+ *           2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 17, 18, 17, 18, 17, 18
+ *
+ * Final column, doing gather([0, 2])
+ *
+ * {
+ *   {
+ *      {2, 3}, {4, 5}
+ *   },
+ *   {
+ *      {15, 16}, {17, 18}, {17, 18}, {17, 18}, {17, 18}
+ *   }
+ * }
+ *
+ * List<List<int32_t>>:
+ * Length : 2
+ * Offsets : 0, 2, 7
+ * Children :
+ *    List<int32_t>:
+ *    Length : 7
+ *    Offsets : 0, 2, 4, 6, 8, 10, 12, 14
+ *       Children :
+ *          2, 3, 4, 5, 15, 16, 17, 18, 17, 18, 17, 18, 17, 18
+ * @endcode
+ *
+ * @tparam MapItType Iterator type to access the incoming column.
+ * @param source_column View into the column to gather from
+ * @param gather_map Iterator access to the gather map for `source_column` map
+ * @param gather_map_size Size of the gather map.
+ * @param nullify_out_of_bounds Nullify values in `gather_map` that are out of bounds
+ * @param stream CUDA stream on which to execute kernels
+ * @param mr Memory resource to use for all allocations
+ * @param prev_base_offsets (optional) The buffer backing the base offsets used in the gather map.
+ *
+ * @returns The gather_data struct needed to construct the gather map for the
+ *          next level of recursion.
+ *
+ */
+template <bool NullifyOutOfBounds, typename MapItType>
+gather_data make_gather_data(
+  cudf::lists_column_view const& source_column,
+  MapItType gather_map,
+  size_type gather_map_size,
+  cudaStream_t stream,
+  rmm::mr::device_memory_resource* mr,
+  rmm::device_uvector<size_type> prev_base_offsets = rmm::device_uvector<size_type>{})
+{
+  // size of the gather map is the # of output rows
+  size_type output_count = gather_map_size;
+  size_type offset_count = output_count + 1;
+
+  // offsets of the source column
+  size_type const* src_offsets{source_column.offsets().data<size_type>()};
+  int src_size = source_column.size();
+
+  // outgoing offsets.  these will persist as output from the entire gather operation
+  auto dst_offsets_c = cudf::make_fixed_width_column(
+    data_type{type_id::INT32}, offset_count, mask_state::UNALLOCATED, stream, mr);
+  mutable_column_view dst_offsets_v = dst_offsets_c->mutable_view();
+  size_type* dst_offsets            = dst_offsets_v.data<size_type>();
```

Comment:
You are mixing `size_type` and `int32` type here. This could be dangerous if `size_type` changed to `int64_t`.

You may be able future-proof this with by using `data_type{type_to_id<size_type>()}` in the `make_fixed_width_column` instead of `INT32` but I've not actually check that works myself. Otherwise, you should be using `int32_t` instead of `size_type` for the offsets in case `size_type` changes.